flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/PrioritizedOperatorSubtaskState.java

```diff
@@ -0,0 +1,265 @@
+	 * should attempt to restore.
+	 */
+	@Nonnull
+	public Iterator<StateObjectCollection<KeyedStateHandle>> getPrioritizedRawKeyedState() {
+		return prioritizedRawKeyedState.iterator();
+	}
+
+	// -----------------------------------------------------------------------------------------------------------------
+
+	/**
+	 * Returns the managed operator state from the job manager, which represents the ground truth about what this state
+	 * should represent. This is the alternative with lowest priority.
+	 */
+	@Nonnull
+	public StateObjectCollection<OperatorStateHandle> getJobManagerManagedOperatorState() {
+		return lastElement(prioritizedManagedOperatorState);
+	}
+
+	/**
+	 * Returns the raw operator state from the job manager, which represents the ground truth about what this state
+	 * should represent. This is the alternative with lowest priority.
+	 */
+	@Nonnull
+	public StateObjectCollection<OperatorStateHandle> getJobManagerRawOperatorState() {
+		return lastElement(prioritizedRawOperatorState);
+	}
+
+	/**
+	 * Returns the managed keyed state from the job manager, which represents the ground truth about what this state
+	 * should represent. This is the alternative with lowest priority.
+	 */
+	@Nonnull
+	public StateObjectCollection<KeyedStateHandle> getJobManagerManagedKeyedState() {
+		return lastElement(prioritizedManagedKeyedState);
+	}
+
+	/**
+	 * Returns the raw keyed state from the job manager, which represents the ground truth about what this state
+	 * should represent. This is the alternative with lowest priority.
+	 */
+	@Nonnull
+	public StateObjectCollection<KeyedStateHandle> getJobManagerRawKeyedState() {
+		return lastElement(prioritizedRawKeyedState);
+	}
+
+	// -----------------------------------------------------------------------------------------------------------------
+
+	/**
+	 * Returns true if this was created for a restored operator, false otherwise. Restored operators are operators that
+	 * participated in a previous checkpoint, even if they did not emit any state snapshots.
+	 */
+	public boolean isRestored() {
+		return restored;
+	}
+
+
+	/**
+	 * This helper method resolves the dependencies between the ground truth of the operator state obtained from the
+	 * job manager and potential alternatives for recovery, e.g. from a task-local source.
+	 */
+	protected <T extends StateObject> List<StateObjectCollection<T>> resolvePrioritizedAlternatives(
+		StateObjectCollection<T> jobManagerState,
+		List<StateObjectCollection<T>> alternativesByPriority,
+		BiFunction<T, T, Boolean> approveFun) {
+
+		// Nothing to resolve if there are no alternatives, or the ground truth has already no state, or if we can
+		// assume that a rescaling happened because we find more than one handle in the JM state (this is more a sanity
+		// check).
+		if (alternativesByPriority == null
+			|| alternativesByPriority.isEmpty()
+			|| !jobManagerState.hasState()
+			|| jobManagerState.size() != 1) {
+
```

Comment:
This means that current local recovery doesn't support rescale?

Reply:
Exactly, local recovery is for checkpoints not savepoints. Rescaling from checkpoint/savepoint involves a restart of the job, which means that all local state is cleaned up anyways because the job was canceled.

Reply:
Aha, got it. I wonder if someday flink will support online rescaling.. but it a another topic, ignore me...

Reply:
That might be, but there is still not a fundamental problem with the approach if we change the matching/creation of the effective state handles in the future, once it is required. For the time being, keeping it simple keeps maintenance simpler.

Reply:
Agreed.